From 7728e679e477358997ebb0e17ea4a14cd268cb0f Mon Sep 17 00:00:00 2001
From: Eran Ben-Avi <benavi@marvell.com>
Date: Tue, 17 Sep 2013 21:40:10 +0300
Subject: [PATCH 0991/1825] cpu: alp: add support for L1 and L2 cache prefetch

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit ad0b84231be33cb828b73a77724351e247bc14f6

	- Add support for L1 cache prefetch enable
	- Add support for PL310 L2 cache prefetch
	  related features.

	Both L1 and L2 prefetch features are enabled
	by default for ALP.

Signed-off-by: Eran Ben-Avi <benavi@marvell.com>

Change-Id: I09f25cc586813b5c733a7b9ab9d3ceb9f885d8c7
Reviewed-on: http://vgitil04.il.marvell.com:8080/3514
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Eran Ben-Avi <benavi@marvell.com>
Tested-by: Eran Ben-Avi <benavi@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 arch/arm/mm/Kconfig      |   20 ++++++++++++++++++++
 arch/arm/mm/cache-l2x0.c |    8 ++++++++
 arch/arm/mm/proc-v7.S    |    3 +++
 3 files changed, 31 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mm/Kconfig b/arch/arm/mm/Kconfig
index 445ea72..fa19dc8 100644
--- a/arch/arm/mm/Kconfig
+++ b/arch/arm/mm/Kconfig
@@ -801,6 +801,26 @@ config CPU_BPREDICT_DISABLE
 	help
 	  Say Y here to disable branch prediction.  If unsure, say N.
 
+config CPU_L1_CACHE_PREF_ENABLE
+	bool "Enable CPU L1 Cache Prefetch"
+	depends on CPU_V7
+	default y if CPU_V7 && ARCH_AVANTA_LP
+	help
+	  Say Y here to enable L1 cache prefetch.
+	  If unsure, say N.
+	  L1 cache prefetch is performance feature,
+	  impact mainly cpu read operations.
+
+config PL310_CACHE_PREF_ENABLE
+	bool "Enable PL310 L2 Cache Prefetch"
+	depends on CACHE_PL310
+	default y if CACHE_PL310 && ARCH_AVANTA_LP
+	help
+	  Say Y here to enable PL310 L2 cache prefetch features.
+	  If unsure, say N.
+	  PL310 L2 cache prefetch-related features may improve
+	  overall system performance.
+
 config TLS_REG_EMUL
 	bool
 	help
diff --git a/arch/arm/mm/cache-l2x0.c b/arch/arm/mm/cache-l2x0.c
index 2a8e380..6683e26 100644
--- a/arch/arm/mm/cache-l2x0.c
+++ b/arch/arm/mm/cache-l2x0.c
@@ -372,6 +372,14 @@ void __init l2x0_init(void __iomem *base, u32 aux_val, u32 aux_mask)
 
 		l2x0_inv_all();
 
+#ifdef CONFIG_PL310_CACHE_PREF_ENABLE
+	/* Support following configuration:
+	 *  Incr double linefill enable
+	 *  Data prefetch enable
+	 *  Double linefill enable
+	 */
+	writel_relaxed(0x50800000, l2x0_base + L2X0_PREFETCH_CTRL);
+#endif
 		/* enable L2X0 */
 		writel_relaxed(1, l2x0_base + L2X0_CTRL);
 	}
diff --git a/arch/arm/mm/proc-v7.S b/arch/arm/mm/proc-v7.S
index 64dbf5b..0a09c8b 100644
--- a/arch/arm/mm/proc-v7.S
+++ b/arch/arm/mm/proc-v7.S
@@ -162,6 +162,9 @@ __v7_ca9mp_setup:
 	mcr     p15, 0, r0, c9, c14, 0
 #endif
 	mov	r10, #(1 << 0)			@ TLB ops broadcasting
+#ifdef CONFIG_CPU_L1_CACHE_PREF_ENABLE
+	orr	r10, r10, #(1 << 2)
+#endif
 	b	1f
 __v7_ca7mp_setup:
 __v7_ca15mp_setup:
-- 
1.7.5.4

